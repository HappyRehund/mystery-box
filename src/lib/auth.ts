// lib/auth.ts
import { getAuth } from '@clerk/nextjs/server';
import prisma from './prisma';
import { NextRequest } from 'next/server';

export async function getCurrentUser(req: NextRequest) {
  const { userId } = getAuth(req);
  
  if (!userId) {
    return null;
  }

  return await prisma.user.findUnique({
    where: {
      id: userId,
    },
  });
}

export async function requireAuth(req: NextRequest) {
  const user = await getCurrentUser(req);
  
  if (!user) {
    throw new Error('Unauthorized');
  }
  
  return user;
}