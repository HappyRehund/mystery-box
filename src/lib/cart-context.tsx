'use client';

import React, { createContext, useContext, useState, useEffect, ReactNode } from 'react';
import { toast } from 'sonner';

export interface CartItem {
  id: string;
  mysteryBoxId: string;
  name: string;
  price: number;
  quantity: number;
  imageUrl: string;
}

interface CartContextType {
  items: CartItem[];
  orderId: string | null;
  isLoading: boolean;
  addItem: (mysteryBoxId: string) => Promise<void>;
  updateQuantity: (itemId: string, quantity: number) => Promise<void>;
  removeItem: (itemId: string) => Promise<void>;
  subtotal: number;
  itemCount: number;
}

const CartContext = createContext<CartContextType | undefined>(undefined);

export function CartProvider({ children }: { children: ReactNode }) {
  const [items, setItems] = useState<CartItem[]>([]);
  const [orderId, setOrderId] = useState<string | null>(null);
  const [isLoading, setIsLoading] = useState(true);

  const subtotal = items.reduce((total, item) => total + item.price * item.quantity, 0);
  const itemCount = items.reduce((count, item) => count + item.quantity, 0);

  // Initialize cart on client-side
  useEffect(() => {
    const initCart = async () => {
      try {
        // Check for existing cart/order in localStorage
        const storedOrderId = localStorage.getItem('orderId');
        
        if (storedOrderId) {
          try {
            // Try to fetch the existing order
            const response = await fetch(`/api/orders/${storedOrderId}`);
            
            if (response.ok) {
              const order = await response.json();
              setOrderId(order.id);
              
              // Transform order items to CartItems
              const cartItems: CartItem[] = order.items.map((item: any) => ({
                id: item.id,
                mysteryBoxId: item.mysteryBoxId,
                name: item.mysteryBox.name,
                price: parseFloat(item.price.toString()),
                quantity: item.quantity,
                imageUrl: item.mysteryBox.imageUrl || '/placeholder.png',
              }));
              
              setItems(cartItems);
            } else {
              // If order not found or not accessible, create new order
              await createNewOrder();
            }
          } catch (error) {
            console.error("Error fetching cart:", error);
            await createNewOrder();
          }
        } else {
          // No order ID in storage, create new order
          await createNewOrder();
        }
      } catch (error) {
        console.error("Error initializing cart:", error);
        toast.error("Failed to load your cart. Please try again later.");
      } finally {
        setIsLoading(false);
      }
    };

    initCart();
  }, []);

  const createNewOrder = async () => {
    try {
      const response = await fetch('/api/orders', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
      });

      if (response.ok) {
        const newOrder = await response.json();
        setOrderId(newOrder.id);
        localStorage.setItem('orderId', newOrder.id);
        setItems([]);
      } else {
        console.error("Failed to create order:", await response.text());
        // Handle guest cart in localStorage if API fails
        setItems([]);
      }
    } catch (error) {
      console.error("Error creating order:", error);
      // Handle guest cart in localStorage if API fails
      setItems([]);
    }
  };

  const addItem = async (mysteryBoxId: string) => {
    if (!orderId) {
      await createNewOrder();
      toast.error("Could not add item. Please try again.");
      return;
    }

    setIsLoading(true);
    try {
      // Fetch mystery box details
      const boxResponse = await fetch(`/api/mystery-boxes/${mysteryBoxId}`);
      if (!boxResponse.ok) {
        throw new Error("Failed to fetch mystery box details");
      }
      const mysteryBox = await boxResponse.json();

      // Check if item already exists in cart
      const existingItem = items.find(item => item.mysteryBoxId === mysteryBoxId);
      
      if (existingItem) {
        // If it exists, update quantity instead
        await updateQuantity(existingItem.id, existingItem.quantity + 1);
        return;
      }

      // Add item to order
      const response = await fetch(`/api/orders/${orderId}/items`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({
          mysteryBoxId,
          quantity: 1,
        }),
      });

      if (response.ok) {
        const newItem = await response.json();
        
        // Add item to local state
        const cartItem: CartItem = {
          id: newItem.id,
          mysteryBoxId: mysteryBox.id,
          name: mysteryBox.name,
          price: parseFloat(mysteryBox.price.toString()),
          quantity: 1,
          imageUrl: mysteryBox.imageUrl || '/placeholder.png',
        };
        
        setItems([...items, cartItem]);
        toast.success(`${mysteryBox.name} added to cart`);
      } else {
        throw new Error("Failed to add item to cart");
      }
    } catch (error) {
      console.error("Error adding item to cart:", error);
      toast.error("Failed to add item to cart. Please try again.");
    } finally {
      setIsLoading(false);
    }
  };

  const updateQuantity = async (itemId: string, quantity: number) => {
    if (quantity < 1) return;
    
    try {
      // For now, we'll just update locally
      setItems(items.map(item => 
        item.id === itemId ? { ...item, quantity } : item
      ));
      
      // API call in a real app:
      // await fetch(`/api/orders/${orderId}/items/${itemId}`, {
      //   method: 'PATCH',
      //   headers: {
      //     'Content-Type': 'application/json',
      //   },
      //   body: JSON.stringify({ quantity }),
      // });
      
      toast.success("Cart updated");
    } catch (error) {
      console.error("Error updating quantity:", error);
      toast.error("Failed to update cart. Please try again.");
    }
  };

  const removeItem = async (itemId: string) => {
    try {
      // For now, we'll just remove locally
      const itemToRemove = items.find(item => item.id === itemId);
      setItems(items.filter(item => item.id !== itemId));
      
      // API call in a real app:
      // await fetch(`/api/orders/${orderId}/items/${itemId}`, {
      //   method: 'DELETE',
      // });
      
      if (itemToRemove) {
        toast.success(`${itemToRemove.name} removed from cart`);
      }
    } catch (error) {
      console.error("Error removing item:", error);
      toast.error("Failed to remove item. Please try again.");
    }
  };

  return (
    <CartContext.Provider
      value={{
        items,
        orderId,
        isLoading,
        addItem,
        updateQuantity,
        removeItem,
        subtotal,
        itemCount,
      }}
    >
      {children}
    </CartContext.Provider>
  );
}

export function useCart() {
  const context = useContext(CartContext);
  if (context === undefined) {
    throw new Error('useCart must be used within a CartProvider');
  }
  return context;
}