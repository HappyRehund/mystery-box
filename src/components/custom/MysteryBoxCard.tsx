import Image from "next/image";
import Link from "next/link";
import { Card, CardContent, CardFooter } from "@/components/ui/card";
import { Button } from "@/components/ui/button";
import { Badge } from "@/components/ui/badge";
import { formatCurrency } from "@/lib/utils";

interface MysteryBoxCardProps {
  id: string;
  name: string;
  description?: string;
  price: number;
  imageUrl?: string;
  categoryName?: string;
  onAddToCart: (id: string) => void;
}

export default function MysteryBoxCard({
  id,
  name,
  description,
  price,
  imageUrl = "/placeholder.png",
  categoryName,
  onAddToCart,
}: MysteryBoxCardProps) {
  return (
    <Card className="overflow-hidden h-full flex flex-col">
      <Link href={`/mystery-box/${id}`} className="relative h-48 w-full">
        <Image
          src={imageUrl}
          alt={name}
          fill
          className="object-cover"
        />
      </Link>
      <CardContent className="p-4 flex-grow">
        <div className="flex justify-between items-start mb-2">
          <Link href={`/mystery-box/${id}`}>
            <h3 className="font-semibold text-foreground hover:text-primary">{name}</h3>
          </Link>
          <span className="font-bold text-primary">{formatCurrency(price)}</span>
        </div>
        {categoryName && (
          <Badge variant="secondary" className="mb-2">
            {categoryName}
          </Badge>
        )}
        {description && (
          <p className="text-sm text-muted-foreground mt-1 line-clamp-3">{description}</p>
        )}
      </CardContent>
      <CardFooter className="p-4 pt-0">
        <Button
          onClick={() => onAddToCart(id)}
          className="w-full"
          variant="default"
        >
          Add to Cart
        </Button>
      </CardFooter>
    </Card>
  );
}