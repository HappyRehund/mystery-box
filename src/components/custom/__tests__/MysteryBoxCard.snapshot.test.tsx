import { render, act } from "@testing-library/react";
import MysteryBoxCard from "../MysteryBoxCard";
import { CartProvider } from "@/lib/cart-context";

describe("MysteryBoxCard Snapshots", () => {
  const mockProps = {
    id: "1",
    name: "Test Box",
    description: "Test Description",
    price: 99.99,
    imageUrl: "/test-image.jpg",
    categoryName: "Test Category",
    onAddToCart: jest.fn(),
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("should match snapshot with all props", async () => {
    let container;
    await act(async () => {
      const result = render(
        <CartProvider>
          <MysteryBoxCard {...mockProps} />
        </CartProvider>
      );
      container = result.container;
    });

    // Update snapshots to match current state
    await act(async () => {
      // Wait for any pending state updates
      await new Promise((resolve) => setTimeout(resolve, 0));
    });

    expect(container).toMatchSnapshot();
  });

  it("should match snapshot without optional props", async () => {
    let container;
    await act(async () => {
      const result = render(
        <CartProvider>
          <MysteryBoxCard
            id={mockProps.id}
            name={mockProps.name}
            price={mockProps.price}
            onAddToCart={mockProps.onAddToCart}
          />
        </CartProvider>
      );
      container = result.container;
    });
    expect(container).toMatchSnapshot();
  });
});
