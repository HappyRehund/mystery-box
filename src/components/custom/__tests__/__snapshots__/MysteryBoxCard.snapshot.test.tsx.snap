// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MysteryBoxCard Snapshots should match snapshot with all props 1`] = `
<div>
  <div
    class="bg-card text-card-foreground gap-6 rounded-xl border py-6 shadow-sm overflow-hidden h-full flex flex-col"
    data-slot="card"
  >
    <a
      class="relative h-48 w-full"
      href="/mystery-box/1"
    >
      <img
        alt="Test Box"
        class="object-cover"
        data-nimg="fill"
        decoding="async"
        loading="lazy"
        sizes="100vw"
        src="/_next/image?url=%2Ftest-image.jpg&w=3840&q=75"
        srcset="/_next/image?url=%2Ftest-image.jpg&w=640&q=75 640w, /_next/image?url=%2Ftest-image.jpg&w=750&q=75 750w, /_next/image?url=%2Ftest-image.jpg&w=828&q=75 828w, /_next/image?url=%2Ftest-image.jpg&w=1080&q=75 1080w, /_next/image?url=%2Ftest-image.jpg&w=1200&q=75 1200w, /_next/image?url=%2Ftest-image.jpg&w=1920&q=75 1920w, /_next/image?url=%2Ftest-image.jpg&w=2048&q=75 2048w, /_next/image?url=%2Ftest-image.jpg&w=3840&q=75 3840w"
        style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
      />
    </a>
    <div
      class="p-4 flex-grow"
      data-slot="card-content"
    >
      <div
        class="flex justify-between items-start mb-2"
      >
        <a
          href="/mystery-box/1"
        >
          <h3
            class="font-semibold text-foreground hover:text-primary"
          >
            Test Box
          </h3>
        </a>
        <span
          class="font-bold text-primary"
        >
          Rp 99,99
        </span>
      </div>
      <span
        class="inline-flex items-center justify-center rounded-md border px-2 py-0.5 text-xs font-medium w-fit whitespace-nowrap shrink-0 [&>svg]:size-3 gap-1 [&>svg]:pointer-events-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive transition-[color,box-shadow] overflow-hidden border-transparent bg-secondary text-secondary-foreground [a&]:hover:bg-secondary/90 mb-2"
        data-slot="badge"
      >
        Test Category
      </span>
      <p
        class="text-sm text-muted-foreground mt-1 line-clamp-3"
      >
        Test Description
      </p>
    </div>
    <div
      class="flex items-center [.border-t]:pt-6 p-4 pt-0"
      data-slot="card-footer"
    >
      <button
        class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive bg-primary text-primary-foreground shadow-xs hover:bg-primary/90 h-9 px-4 py-2 has-[>svg]:px-3 w-full"
        data-slot="button"
      >
        Add to Cart
      </button>
    </div>
  </div>
</div>
`;

exports[`MysteryBoxCard Snapshots should match snapshot without optional props 1`] = `
<div>
  <div
    class="bg-card text-card-foreground gap-6 rounded-xl border py-6 shadow-sm overflow-hidden h-full flex flex-col"
    data-slot="card"
  >
    <a
      class="relative h-48 w-full"
      href="/mystery-box/1"
    >
      <img
        alt="Test Box"
        class="object-cover"
        data-nimg="fill"
        decoding="async"
        loading="lazy"
        sizes="100vw"
        src="/_next/image?url=%2Fimages%2Ftablet-box.webp&w=3840&q=75"
        srcset="/_next/image?url=%2Fimages%2Ftablet-box.webp&w=640&q=75 640w, /_next/image?url=%2Fimages%2Ftablet-box.webp&w=750&q=75 750w, /_next/image?url=%2Fimages%2Ftablet-box.webp&w=828&q=75 828w, /_next/image?url=%2Fimages%2Ftablet-box.webp&w=1080&q=75 1080w, /_next/image?url=%2Fimages%2Ftablet-box.webp&w=1200&q=75 1200w, /_next/image?url=%2Fimages%2Ftablet-box.webp&w=1920&q=75 1920w, /_next/image?url=%2Fimages%2Ftablet-box.webp&w=2048&q=75 2048w, /_next/image?url=%2Fimages%2Ftablet-box.webp&w=3840&q=75 3840w"
        style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
      />
    </a>
    <div
      class="p-4 flex-grow"
      data-slot="card-content"
    >
      <div
        class="flex justify-between items-start mb-2"
      >
        <a
          href="/mystery-box/1"
        >
          <h3
            class="font-semibold text-foreground hover:text-primary"
          >
            Test Box
          </h3>
        </a>
        <span
          class="font-bold text-primary"
        >
          Rp 99,99
        </span>
      </div>
    </div>
    <div
      class="flex items-center [.border-t]:pt-6 p-4 pt-0"
      data-slot="card-footer"
    >
      <button
        class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive bg-primary text-primary-foreground shadow-xs hover:bg-primary/90 h-9 px-4 py-2 has-[>svg]:px-3 w-full"
        data-slot="button"
      >
        Add to Cart
      </button>
    </div>
  </div>
</div>
`;
